crates/ty_python_semantic/resources/mdtest/import/dunder_all.md

```diff
@@ -0,0 +1,796 @@
+
+```py
+__all__ = ("X",)
+
+class X: ...
+
+__all__: tuple[str, ...] = ("Y", "Z")
+
+class Y: ...
+class Z: ...
+```
+
+`importer.py`:
+
+```py
+import exporter
+import exporter_annotated
+from ty_extensions import dunder_all_names
+
+# revealed: tuple[Literal["C"], Literal["D"]]
+reveal_type(dunder_all_names(exporter))
+
+# revealed: tuple[Literal["Y"], Literal["Z"]]
+reveal_type(dunder_all_names(exporter_annotated))
+```
+
+### Augmenting list with a list or submodule `__all__`
+
+`subexporter.py`:
+
+```py
+__all__ = ["A", "B"]
+
+class A: ...
+class B: ...
+```
+
+`exporter.py`:
+
+```py
+import subexporter
+
+__all__ = []
+__all__ += ["C", "D"]
+__all__ += subexporter.__all__
+
+class C: ...
+class D: ...
+```
+
+`importer.py`:
+
+```py
+import exporter
+from ty_extensions import dunder_all_names
+
+# revealed: tuple[Literal["A"], Literal["B"], Literal["C"], Literal["D"]]
+reveal_type(dunder_all_names(exporter))
+```
+
+### Extending with a list or submodule `__all__`
+
+`subexporter.py`:
+
+```py
+__all__ = ["A", "B"]
+
+class A: ...
+class B: ...
+```
+
+`exporter.py`:
+
+```py
+import subexporter
+
+__all__ = []
+__all__.extend(["C", "D"])
+__all__.extend(subexporter.__all__)
+
+class C: ...
+class D: ...
+```
+
+`importer.py`:
+
+```py
+import exporter
+from ty_extensions import dunder_all_names
+
+# revealed: tuple[Literal["A"], Literal["B"], Literal["C"], Literal["D"]]
+reveal_type(dunder_all_names(exporter))
+```
+
+### Appending a single symbol
+
+`exporter.py`:
+
+```py
+__all__ = ["A"]
+__all__.append("B")
+
+class A: ...
+class B: ...
+```
+
+`importer.py`:
+
+```py
+import exporter
+from ty_extensions import dunder_all_names
+
+# revealed: tuple[Literal["A"], Literal["B"]]
+reveal_type(dunder_all_names(exporter))
+```
+
+### Removing a single symbol
+
+`exporter.py`:
+
+```py
+__all__ = ["A", "B"]
+__all__.remove("A")
+
+# Non-existant symbol in `__all__` at this point
+# TODO: This raises `ValueError` at runtime, maybe we should raise a diagnostic as well?
+__all__.remove("C")
+
+class A: ...
+class B: ...
+```
+
+`importer.py`:
+
+```py
+import exporter
+from ty_extensions import dunder_all_names
+
+# revealed: tuple[Literal["B"]]
+reveal_type(dunder_all_names(exporter))
+```
+
+### Mixed
+
+`subexporter.py`:
+
+```py
+__all__ = []
+
+__all__ = ["A"]
+__all__.append("B")
+__all__.extend(["C"])
+__all__.remove("B")
+
+class A: ...
+class B: ...
+class C: ...
+```
+
+`exporter.py`:
+
+```py
+import subexporter
+
+__all__ = []
+__all__ += ["D"]
+__all__ += subexporter.__all__
+
+class D: ...
+```
+
+`importer.py`:
+
+```py
+import exporter
+from ty_extensions import dunder_all_names
+
+# revealed: tuple[Literal["A"], Literal["C"], Literal["D"]]
+reveal_type(dunder_all_names(exporter))
+```
+
+## Invalid
+
+### Unsupported idioms
+
+Idioms that are not mentioned in the [specification] are not recognized by `ty` and if they're used,
+`__all__` is considered to be undefined for that module. This is to avoid false positives.
```

Comment:
I guess the case where we are importing a non-re-export, which is listed in the dynamic `__all__`, doesn't  count as a false positive that we need to care about, since a non-re-export can only occur in a stub file, and in a stub file there is no reason to ever use dynamic `__all__` construction that type checkers don't understand.